edisgo/network/topology.py

```diff
@@ -2399,61 +2455,199 @@ def connect_to_lv_based_on_geolocation(
+            comp_type_counts = (
+                comp_df.loc[comp_df.bus.isin(lv_buses_masked.index)]
+                .groupby("bus")
+                .size()
+            )
+            lv_buses_masked.loc[:, "num_comps"] = (
+                lv_buses_masked.index.map(comp_type_counts).fillna(0).astype(int)
+            )
+            # get buses where the least number of components of the given type is
+            # connected
+            lv_buses_masked = lv_buses_masked[
+                lv_buses_masked.num_comps == lv_buses_masked.num_comps.min()
+            ]
+
+            if lv_buses_masked.num_comps.min() >= allowed_number_of_comp_per_bus:
+                # if all buses within the allowed distance have equal or more
+                # components of the same type connected to them than allowed,
+                # connect to new bus
+                target_bus = self._connect_to_lv_bus(
+                    edisgo_object, lv_buses.distance.idxmin(), comp_type, comp_data
+                )
+                lv_buses_masked = pd.DataFrame(self.buses_df.loc[target_bus]).T
+                lv_buses_masked["distance"] = 0
+                mv_buses_masked = pd.DataFrame()
+
+            target_bus = lv_buses_masked.loc[lv_buses_masked.distance.idxmin()]
+            if isinstance(target_bus, pd.DataFrame):
+                target_bus = target_bus.iloc[0]
+            return target_bus.name
+
+        # Ensure 'p' is in comp_data, defaulting to 'p_set' or 'p_nom'
         if "p" not in comp_data.keys():
             comp_data["p"] = (
                 comp_data["p_set"]
                 if "p_set" in comp_data.keys()
                 else comp_data["p_nom"]
             )
 
-        voltage_level = comp_data.pop("voltage_level")
-        if voltage_level not in [6, 7]:
-            raise ValueError(
-                f"Voltage level must either be 6 or 7 but given voltage level "
-                f"is {voltage_level}."
-            )
+        # Extract and validate voltage level
+        voltage_level = comp_data.get("voltage_level")
+        validate_voltage_level(voltage_level)
         geolocation = comp_data.get("geom")
 
-        if comp_type == "generator":
-            add_func = self.add_generator
-        elif comp_type == "charging_point" or comp_type == "heat_pump":
-            add_func = self.add_load
+        # Determine the appropriate add function based on component type
+        add_func = get_add_function(comp_type)
+
+        # Set the component type in comp_data if necessary
+        if comp_type in ["charging_point", "heat_pump"]:
             comp_data["type"] = comp_type
-        elif comp_type == "storage_unit":
-            add_func = self.add_storage_unit
-        else:
-            logger.error(f"Component type {comp_type} is not a valid option.")
-            return
+        elif comp_type in ["generator", "storage_unit"]:
+            comp_data["p_nom"] = comp_data["p"]
```

Comment:
Why do you need this? Above, p is set, so that all have p. Setting both seems redundant.